pipeline/relation_embedder/relation_embedder/src/main/scala/uk/ac/wellcome/relation_embedder/ArchiveRelationsCache.scala

```diff
@@ -74,18 +72,43 @@ class ArchiveRelationsCache(
         getAncestors(parentPath, relations(parentPath) :: accum)
     }
 
+  private def getNumChildren(path: String): Int =
+    childMapping.get(path).map(_.length).getOrElse(0)
+
+  private def getNumDescendents(path: String): Int = {
+    @tailrec
+    def numDescendents(stack: List[String], accum: Int = 0): Int =
```

Comment:
I don't believe this needs to be a closure, it doesn't close over `path`

Reply:
It's not a closure, just a function defined within another to hide it from the outside.

Reply:
Exactly, it's not capturing anything - so you're redefining the inner function every time `getNumDescendents` is called

Reply:
Hm not sure whether the scala compiler actually does that under the hood, and if so I am more bothered with keeping in organised and hidden than premature optimisation for this case

Reply:
I don't know about Scala compile but for Rust for example a closure function will be compiled to a single function which takes an additional struct containing the captured environment.

Reply:
I'm talking more about readability/intent than performance - I spent a while looking for `path` in the inner function because we usually only use inner functions to close over something. It's a nit regardless...

Reply:
Ah ok. Personally I don't think expect a closure whenever I see an inner function, the official scala docs has a very similar example for defining a recursive function: https://docs.scala-lang.org/tour/nested-functions.html